### PR TITLE
Add gateway timeout to pubsub pull

### PR DESCRIPTION
**Pull request by daspecster**
Based from #2575.

Implement exception for pubsub subscription pull.

Fixes: #2574

### CODE REVIEW

**Comment by dhermes** [comment]
Ref: https://github.com/googleapis/googleapis/blob/37cc0e5acae50ee91f00827a7010c3b07dfa5311/google/rpc/code.proto#L69-L70

**Comment by dhermes** [comment]
Relates to my [comment][1] on #2496

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2496#issuecomment-251503719

**Comment by tseaver** [comment]
Rather than raising a 504, wouldn't we rather match the semantics of the JSON API and return an empty list?

**Comment by daspecster** [comment]
@tseaver could do that that. How would the user know to retry in that case?

**Comment by dhermes** [comment]
@tseaver I am :-1: on swallowing exceptions. It's up to the API implementers to make sure the JSON and gRPC surface is sufficiently similar.

**Comment by tseaver** [comment]
@daspecster 

> How would the user know to retry in that case?

A timeout exception due to no messages isn't really "exceptional." Long-running subscriber processes are going to be retrying anyway, inside a `while True` sort of loop.  Short-lived subscribers just iterate over the (possibly empty) list of messages and exit.

**Comment by daspecster** [comment]
@tseaver, wouldn't they `while True:` continue after the `DEADLINE_EXCEEDED` in that case?

Or I guess I would think that if the deadline was exceeded then the user would need to take action on that information?

**Comment by tseaver** [comment]
@dhermes 

>  It's up to the API implementers to make sure the JSON and gRPC surface is sufficiently similar.

We paper over *lots* of differences between the different API surfaces already:  it has been an explicit goal that the user doesn't notice or care about them.

@daspecster I'm arguing that users shouldn't need to care about this gRPC-only timeout exception at all:  if we propagate it, then they have to catch it, but it doesn't "mean" anything to them beyond "there are no messages to process," which is exactly what returning an empty sequence tells them.

**Comment by daspecster** [comment]
>A timeout exception due to no messages isn't really "exceptional." 

@tseaver I'm not seeing where this is the same or different from the ackDeadline or other normal operations.

It sounds like if there were no messages after the deadline then it would just return nothing and not throw an error?

>Received Pub/Sub messages. The Pub/Sub system will return zero messages if there are no more available in the backlog. The Pub/Sub system may return fewer than the maxMessages requested even if there are more messages available in the backlog.

https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#pullresponse

**Comment by dhermes** [comment]
> We paper over lots of differences between the different API surfaces already

@tseaver I'm not disputing that but this is an error being converted into a successful (empty) response. We don't even have a repro yet confirming this is intended / deterministic behavior. We should at least hear from someone on the API team / see official docs that confirm a gRPC 504 is equivalent to an empty result.

**Comment by tseaver** [comment]
> I'm not seeing where this is the same or different from the ackDeadline or other normal operations.

`ackDeadline` is a [configuration option for the `Subscription`](https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#google.pubsub.v1.Subscription):  pull-mode subscribers are required to `ack` pulled messages within that interval, or else they get queued for redelivery.  It can't be relevant to #2574, as the `DEADLINE_EXCEEED` message is being raised *inside* `subscription.pull()`.

**Comment by tseaver** [comment]
> We should at least hear from someone on the API team / see official docs that confirm a gRPC 504 is equivalent to an empty result.

@tmatsuo Can you chime in here?

**Comment by daspecster** [comment]
@tmatsuo could you check this out when you have a second?

**Comment by daspecster** [comment]
@dhermes @tseaver, What shall I do with this?

**Comment by daspecster** [comment]
Bump

**Comment by dhermes** [comment]
@daspecster Can you give snippets / stacktraces of what happens for the exact same client code, but one with gRPC disabled and one with it enabled?

**Comment by daspecster** [comment]
@dhermes @tseaver.


Here is my test file.
```python
from google.cloud import pubsub


client = pubsub.Client()

topic = client.topic('test-padding')
if not topic.exists():
    topic.create()

subscription = topic.subscription('test-padding')

if not subscription.exists():
    subscription.create()

messages = subscription.pull()
print(messages)
```

With gRPC:
```bash
$ python test_pubsub_padding.py
Traceback (most recent call last):
  File "test_pubsub_padding.py", line 18, in <module>
    messages = subscription.pull()
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/cloud/pubsub/subscription.py", line 328, in pull
    self.full_name, return_immediately, max_messages)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/cloud/pubsub/_gax.py", line 414, in subscription_pull
    return_immediately=return_immediately)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/cloud/gapic/pubsub/v1/subscriber_api.py", line 545, in pull
    return self._pull(request, options)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/gax/api_callable.py", line 481, in inner
    return api_caller(api_call, this_settings, request)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/gax/api_callable.py", line 469, in base_caller
    return api_call(*args)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/gax/api_callable.py", line 434, in inner
    errors.create_error('RPC failed', cause=exception))
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/gax/api_callable.py", line 430, in inner
    return a_func(*args, **kwargs)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/google/gax/api_callable.py", line 64, in inner
    return a_func(*updated_args, **kwargs)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/grpc/_channel.py", line 481, in __call__
    return _end_unary_response_blocking(state, False, deadline)
  File "/Users/daspecster/.virtualenvs/gct/lib/python2.7/site-packages/grpc/_channel.py", line 432, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
google.gax.errors.GaxError: GaxError(RPC failed, caused by <_Rendezvous of RPC that terminated with (StatusCode.DEADLINE_EXCEEDED, Deadline Exceeded)>)
E1121 11:08:01.032859000 140736417121216 network_status_tracker.c:48] Memory leaked as all network endpoints were not shut down
```


With gRPC disabled:
```bash
$ GOOGLE_CLOUD_DISABLE_GRPC=true python test_pubsub_padding.py
[]
```

**Comment by dhermes** [comment]
I didn't want it to be true, so I stripped out our library and just made low(ish)-level requests with each transport: https://gist.github.com/dhermes/ac47b7b6eebfd46febbd9ee3d824436d

Unfortunately it **IS** true. The HTTP surface returns a 200 and the gRPC surface returns with an error set to `DEADLINE_EXCEEDED`.

After confirming this sad "contradiction", I now agree with @tseaver that we should wrap the `DEADLINE_EXCEEDED` and just return whatever the equivalent to an empty result set is.

**Comment by dhermes** [comment]
60s to wait for a timeout is a long time, but I feel like if we do make this assumption (i.e. that the surfaces differ), then we should bake that assumption into the system tests, so we can quickly fix it if the assumption ceases to be true.

File: pubsub/google/cloud/pubsub/_gax.py [hunk 1]
```diff
@@ -415,6 +415,8 @@ def subscription_pull(self, subscription_path, return_immediately=False,
         except GaxError as exc:
             if exc_to_code(exc.cause) == StatusCode.NOT_FOUND:
                 raise NotFound(subscription_path)
+            if exc_to_code(exc.cause) == StatusCode.DEADLINE_EXCEEDED:
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub/_gax.py]: Re-factor:

```python
code = exc_to_code(exc.cause)
if code == StatusCode.NOT_FOUND
    raise NotFound(subscription_path)
elif code == StatusCode.DEADLINE_EXCEEDED
    # NOTE: The JSON-over-HTTP API returns a 200 with an empty response
    #       when ``return_immediately`` is ``False``, so we "mutate" the
    #       gRPC error into a non-error to conform.
    if not return_immediately:
        return []

raise
```

Comment (daspecster) [inline pubsub/google/cloud/pubsub/_gax.py]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2576/commits/903c29e8b2af758638863c55b12ce19a47d38d01

**Comment by daspecster** [comment]
Squashed!

File: pubsub/google/cloud/pubsub/_gax.py [hunk 2]
```diff
@@ -413,8 +413,15 @@ def subscription_pull(self, subscription_path, return_immediately=False,
                 subscription_path, max_messages,
                 return_immediately=return_immediately)
         except GaxError as exc:
-            if exc_to_code(exc.cause) == StatusCode.NOT_FOUND:
+            code = exc_to_code(exc.cause)
+            if code == StatusCode.NOT_FOUND:
                 raise NotFound(subscription_path)
+            elif exc_to_code(exc.cause) == StatusCode.DEADLINE_EXCEEDED:
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub/_gax.py]: @daspecster You have repeated the "computation". That's why `code` is created before you start checking

Comment (daspecster) [inline pubsub/google/cloud/pubsub/_gax.py]: Yeah, I don't know how I missed that.

Comment (daspecster) [inline pubsub/google/cloud/pubsub/_gax.py]: changed to 
```python
code = exc_to_code(exc.cause)
            if code == StatusCode.NOT_FOUND:
                raise NotFound(subscription_path)
            elif code == StatusCode.DEADLINE_EXCEEDED:
                # NOTE: The JSON-over-HTTP API returns a 200 with an empty
                #       response when ``return_immediately`` is ``False``, so
                #       we "mutate" the gRPC error into a non-error to conform.
                if not return_immediately:
                    return []
```

and squashed again.

**Pull request merged**